### PR TITLE
[FSDP][StateDict] Allow FULL_STATE_DICT option for 2D (#120837)

### LINKED ISSUE #120722: Allow Full State Dict with 2D FSDP + TP
### 🐛 Describe the bug

Torch does not allow 2D FSDP + TP to get FULL_STATE_DICT. However, if I remove checks here:
https://github.com/pytorch/pytorch/blob/3f62b05d31d4b29d60874b05adc0e5aedbad3722/torch/distributed/fsdp/_state_dict_utils.py#L303-L310
and warnings here:
https://github.com/pytorch/pytorch/blob/3f62b05d31d4b29d60874b05adc0e5aedbad3722/torch/distributed/fsdp/_state_dict_utils.py#L811-L815
it seems to work fine.

Is there a reason we need these checks? Can we remove them given things seem to work fine?

### Versions

Nightly.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### PR DESCRIPTION
**Pull request by mvpatel2000**
Copy of https://github.com/pytorch/pytorch/pull/120837 for release

Fixes #120722

TL;DR for the issue:
As users are expected to use get_model_state_dict to do state_dict retrieval, I think it's fine to remove the warning and RuntimeError. More context in #120722.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/120837
Approved by: https://github.com/Skylion007

### CODE REVIEW

**Comment by mvpatel2000** [comment]
CC: @wz337

**Review by atalman — APPROVED** [review]
